R/cellbrowser.R

```diff
@@ -0,0 +1,499 @@
+#' @param skip.expr.matrix whether to skip exporting expression matrix
+#' @param skip.metadata whether to skip exporting metadata
+#' @param skip.reductions whether to skip exporting reductions
+#' @param port on which port to run UCSC cellbrowser webserver after export
+#' @param ... specifies the metadata fields to export. To supply a field and its
+#' human readable name, pass name as \code{field="name"} parameter.
+#'
+#' @return This function exports Seurat object as a set of tsv files
+#' to \code{dir} directory, copying the \code{markers.file} if it is
+#' passed. It also creates the default \code{cellbrowser.conf} in the
+#' directory. This directory could be read by \code{cbBuild} to
+#' create a static website viewer for the dataset. If \code{cb.dir}
+#' parameter is passed, the function runs \code{cbBuild} (if it is
+#' installed) to create this static website in \code{cb.dir} directory.
+#' If \code{port} parameter is passed, it also runs the webserver for
+#' that directory and opens a browser.
+#'
+#' @author Maximilian Haeussler, Nikolay Markov
+#'
+#' @importFrom tools file_ext
+#' @importFrom utils browseURL
+#' @importFrom reticulate py_module_available import
+#' @importFrom Seurat Project Idents GetAssayData Embeddings FetchData
+#'
+#' @export
+#'
+#' @name CellBrowser
+#' @rdname CellBrowser
+#'
+#' @importFrom methods slot
+#' @importFrom utils packageVersion
+#' @importFrom reticulate py_module_available import
+#'
+#' @examples
+#' \dontrun{
+#' ExportToCellbrowser(pbmc_small, dataset.name = "PBMC", dir = "out")
+#' }
+#'
+ExportToCellbrowser <- function(
+  object,
+  dir,
+  dataset.name = Project(object = object),
+  reductions = NULL,
+  markers.file = NULL,
+  cluster.field = NULL,
+  cb.dir = NULL,
+  port = NULL,
+  use.mtx = FALSE,
+  meta.fields = NULL,
+  meta.fields.names = NULL,
+  matrix.slot = "counts",
+  markers.n = 100,
+  skip.markers = FALSE,
+  skip.expr.matrix = FALSE,
+  skip.metadata = FALSE,
+  skip.reductions = FALSE
+) {
+  if (!requireNamespace("Seurat", quietly = TRUE)) {
+    stop("This script requires that Seurat (V2 or V3) is installed")
+  }
+  message("Seurat Version installed: ", packageVersion(pkg = "Seurat"))
+  object.version <- as.character(x = slot(object = object, name = 'version'))
+  message("Object was created with Seurat version ", object.version)
+  if (substr(x = object.version, start = 1, stop = 1) != '2' && substr(x = object.version, start = 1, stop = 1) != '3') {
+    stop("can only process Seurat2 or Seurat3 objects, version of rds is ", object.version)
+  }
+  if (substr(x = object.version, start = 1, stop = 1) != substr(x = packageVersion(pkg = "Seurat"), start = 1, stop = 1)) {
+    stop("The installed version of Seurat is different from Seurat object loaded. You have to down- or upgrade your installed Seurat version, see the Seurat documentation")
+  }
+  embeddings <- reductions # Seurat prefers the word reductions
```

Comment:
What's the purpose of this line? The cell embeddings are a specific part of the dimensional reduction; however, throughout Seurat and SeuratWrappers, we use "reductions" to refer to a dimensional reduction that we may take parts of (such as the cell embeddings) later

Reply:
Sorry, I never understood the difference and I don't understand your explanation. I changed the variable now to "reducNames", hope this is better.

Reply:
The dimensional reduction is a specific class defined by Seurat [here](https://github.com/satijalab/seurat/blob/master/R/objects.R#L136-L149). The equivalent Python code would be

```python
class DimReduc(object):

    def __init__(self, cell_embeddings, feature_loadings, feature_loadings_projected, assay_used, is_global, stdev, key, jackstraw, misc):
        self.cell_embeddings = cell_embeddings
        self.feature_loadings = feature_loadings
        self.feature_loadings_projected = feature_loadings_projected
        self.assay_used = assay_used
        self.is_global = is_global
        self.stdev = stdev
        self.key = key
        self.jackstraw = jackstraw
        self.misc = misc
```

The cell embeddings are just one matrix in this class

```R
# In R
Embeddings(reduction)
```

```python
# In Python
reduction.cell_embeddings
```

Typically, the reductions are named after the technique run (eg. "pca", "umap", "tsne") and we use this to specify which specific cell embeddings we want

```R
# In R
reduction <- object[["pca"]]
embeddings <- Embeddings(reduction)
# or
embeddings <- Embeddings(object[["pca"]])
```
So it's not wrong to have a parameter named `reductions` to refer to dimensional reductions from which we will pull cell embeddings. For example, [`DimPlot`](https://rdrr.io/cran/Seurat/man/DimPlot.html) takes a parameter `reduction` that takes the name of a dimensional reduction. We use this name to pull cell embedding for visualization later. As such, we accept and encourage the use of the term `reductions` to refer to dimensional reductions that will be used later.

I'm not saying you're doing anything wrong per-say, I'm just trying to explain that we don't prefer the term "embeddings" to refer to a dimensional reduction object from which data will later be pulled